### PR TITLE
codeintel: Remove job handle abstraction

### PR DESCRIPTION
**Pull request by efritz**
This PR removes the codeintel jobHandle class, which is more verbose than necessary to accomplish what it does (a simple wrapper around an identifier).

Summary of changes:

- db.Savepoint now generates an ID (as job handle did)
- db.RollbackToSavepoint now keeps track of valid savepoint identifiers in the current txn
- db.Dequeue now returns a transactional DB instead of a job handle
- db.Dequeue refactored so that adding additional work type s(indexes) will be easier in the near-future
- Updated worker to use new signatures

### CODE REVIEW

File: internal/codeintel/db/dequeue.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/keegancsmith/sqlf"
+	"github.com/pkg/errors"
+)
+
+type dequeueScanner func(rows *sql.Rows, err error) (interface{}, bool, error)
+
+// dequeueRecord selects the record from the given table in a queued state according to the given sort expression.
+// This record is locked in a transaction. The record and the transaction in which it is locked are both returned.
+// This transaction must be closed by the caller. If there is no such unlocked record, a nil record and a nil DB
+// will be returned along with a false-valued flag. This method must not be called from within a transaction.
```
Comment (gbrik) [inline internal/codeintel/db/dequeue.go:16]: question:
why can't we call this from within a transaction?

File: cmd/precise-code-intel-worker/internal/worker/worker.go [hunk 2]
```diff
@@ -83,12 +83,12 @@ func (w *Worker) Stop() {
 func (w *Worker) dequeueAndProcess(ctx context.Context) (_ bool, err error) {
 	start := time.Now()
 
-	upload, jobHandle, ok, err := w.db.Dequeue(ctx)
+	upload, tx, ok, err := w.db.Dequeue(ctx)
```
Comment (gbrik) [inline cmd/precise-code-intel-worker/internal/worker/worker.go:86]: suggestion:
unnecessary for this PR, but we could promote the entire worker instead of just the DB when we transact like:
```suggestion
	upload, w, ok, err := w.Dequeue(ctx)
```
which would make some mistakes impossible to write (e.g. right now you could later erroneously `w.processor.Process(w.db)` instead of `w.processor.Process(tx)`, whereas with the above change the first one would be the only and correct call)

File: internal/codeintel/db/dequeue.go (hunk 1 above)
Comment (efritz) [reply internal/codeintel/db/dequeue.go:16]: The way the record locking works is that it's selected for update/updated outside of a transaction, then a transaction starts where it _actually_ row locks the record.

If we start within a transaction it makes it more difficult to write the same behavior that will allow us to (1) ensure that only one worker is working on a particular row at one time, and (2) allow us to easily determine record which were orphaned by a worker that crashed or was taken down by an orchestrator.

There needs to be some transition of state that's visible to other transactions, and that means you have to start outside of one.

File: cmd/precise-code-intel-worker/internal/worker/worker.go (hunk 2 above)
Comment (efritz) [reply cmd/precise-code-intel-worker/internal/worker/worker.go:86]: That's a good idea. I'll make a ticket for this.

Comment (efritz) [reply cmd/precise-code-intel-worker/internal/worker/worker.go:86]: https://github.com/sourcegraph/sourcegraph/issues/10831

File: internal/codeintel/db/dequeue.go (hunk 1 above)
Comment (gbrik) [reply internal/codeintel/db/dequeue.go:16]: hrmm i feel like i'm missing something: 

if the `UPDATE` query happened in a transaction, wouldn't it lock that row thus preventing other workers from processing it for (1)? don't fully understand (2), do you mean catching rows where state is `processing`? i think these would get rolled back to `queued` on error in that case, so we wouldn't need to worry about them being orphaned?

Comment (efritz) [reply internal/codeintel/db/dequeue.go:16]: Sorry, let me try again:

Outside of a transaction, we do an `UPDATE` that moves a single unlocked upload from queued to processing. This query uses a `SELECT FOR UPDATE SKIP LOCKED` subquery in order to atomatically update a record that isn't a candidate for any other worker.

As this is outside of a transaction, this state transaction is visible to all other processes, so that any other worker trying to update a record will either see that the row is locked (and keep scanning), or it will see that its' no longer in a queued state.

The worker then starts a transaction and performs a `SELECT FOR UPDATE` on the record, which causes a row lock to be held for the rest of the transaction. The worker then updates the state to completed or errored and commits the transaction.

If a worker dies before or inside of the transaction, the upload resetter process (ran in another goroutine of _some_ worker) will see that there's an unlocked row that's been processing for more than 5s (**much** more than the time it takes to update and start a transaction at the beginning for processing). These records can be moved back into the queue state to become eligible for a retry.

Another reason we need to do it this way rather than just relying on row locks is that we need to query the db for all processing/queued/completed/errored records by state. If we relied on `SKIP LOCKED` for these queries we couldn't select any record that's currently processing (efficiently).

Comment (gbrik) [reply internal/codeintel/db/dequeue.go:16]: ok, got it. ty!
